File: python/synapse/planner.py
"""
Planner — converts a DependencyGraph into an ordered execution plan.

The plan is a list of *stages*. Each stage is a group of tool calls
that have no unresolved dependencies between them and can therefore
be launched concurrently.  Stages are computed via Kahn's topological
sort, partitioned by BFS level so that maximum parallelism is achieved
at every step.

Example
-------
Given the dependency graph::

    A ──┐
        ├──► C ──► E
    B ──┘
         D ──────► E (also)

The planner produces::

    Stage 0 (parallel): [A, B, D]
    Stage 1 (parallel): [C]
    Stage 2 (parallel): [E]
"""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field

from .dependency_analyzer import DependencyGraph, ToolCall
from .observability import SynapseLogger


@dataclass
class Stage:
    """A set of tool calls that can be executed concurrently."""

    index: int
    calls: list[ToolCall] = field(default_factory=list)

    def __repr__(self) -> str:
        ids = ", ".join(c.id for c in self.calls)
        return f"Stage({self.index}, [{ids}])"


@dataclass
class ExecutionPlan:
    """
    Ordered list of Stages produced by the Planner.

    Attributes:
        stages:         Ordered list from first to last.
        total_calls:    Total number of tool calls.
        parallelism:    Average number of calls per stage (higher = better).
        critical_path:  Sequence of call ids forming the longest dependency chain.
    """

    stages: list[Stage]
    total_calls: int
    parallelism: float
    critical_path: list[str]

    def __repr__(self) -> str:
        lines = [
            f"ExecutionPlan({self.total_calls} calls, "
            f"{len(self.stages)} stages, "
            f"avg parallelism={self.parallelism:.1f}x, "
            f"critical_path={self.critical_path})"
        ]
        for stage in self.stages:
            ids = [c.id for c in stage.calls]
            marker = " ← critical" if any(i in self.critical_path for i in ids) else ""
            lines.append(f"  Stage {stage.index}: {ids}{marker}")
        return "\n".join(lines)


class Planner:
    """
    Builds an ExecutionPlan from a DependencyGraph using BFS-level
    topological sort (Kahn's algorithm with level tracking).
    """

    def __init__(self, logger: SynapseLogger | None = None) -> None:
        self.logger = logger

    def plan(self, graph: DependencyGraph) -> ExecutionPlan:
        # in_degree: how many unresolved deps each node has
        in_degree: dict[str, int] = {
            cid: len(deps) for cid, deps in graph.edges.items()
        }

        # Start with all roots (zero in-degree)
        queue: deque[str] = deque(
            cid for cid, deg in in_degree.items() if deg == 0
        )

        stages: list[Stage] = []
        processed = 0

        while queue:
            # Everything currently in the queue shares the same BFS level
            level_size = len(queue)
            stage = Stage(index=len(stages))

            for _ in range(level_size):
                cid = queue.popleft()
                stage.calls.append(graph.nodes[cid])
                processed += 1

                # Reduce in-degree for all dependents
                for dependent_id in graph.rev_edges.get(cid, set()):
                    in_degree[dependent_id] -= 1
                    if in_degree[dependent_id] == 0:
                        queue.append(dependent_id)

            stages.append(stage)

        if processed != len(graph.nodes):
            raise ValueError(
                "Cycle detected during planning — not all nodes were processed. "
                f"({processed}/{len(graph.nodes)} processed)"
            )

        total = len(graph.nodes)
        parallelism = total / len(stages) if stages else 0.0
        critical = self._critical_path(graph)
        estimated_speedup = (
            total / len(critical)
            if critical
            else 1.0
        )

        if self.logger is not None:
            self.logger.log(
                "dag_planned",
                parallel_count=max((len(stage.calls) for stage in stages), default=0),
                stage_count=len(stages),
                critical_path=critical,
                estimated_speedup=estimated_speedup,
            )

        return ExecutionPlan(
            stages=stages,
            total_calls=total,
            parallelism=parallelism,
            critical_path=critical,
        )

    # ------------------------------------------------------------------
    # Critical path (longest chain by number of hops)
    # ------------------------------------------------------------------

    def _critical_path(self, graph: DependencyGraph) -> list[str]:
        """
        Returns the sequence of call ids forming the longest dependency
        chain (by hop count).  Ties are broken by id lexicographic order.
        """
        memo: dict[str, list[str]] = {}

        def longest(cid: str) -> list[str]:
            if cid in memo:
                return memo[cid]
            preds = graph.edges.get(cid, set())
            if not preds:
                memo[cid] = [cid]
                return [cid]
            best = max(
                (longest(p) for p in preds),
                key=lambda path: (len(path), path),
            )
            memo[cid] = best + [cid]
            return memo[cid]

        if not graph.nodes:
            return []

        return max(
            (longest(cid) for cid in graph.nodes),
            key=lambda path: (len(path), path),
        )

File: typescript/src/planner.ts
/**
 * Planner — converts a DependencyGraph into an ordered ExecutionPlan.
 *
 * Uses BFS-level topological sort (Kahn's algorithm) to group tool calls
 * into Stages.  All calls within a Stage have no mutual dependencies and
 * can be launched concurrently.
 */

import type { DependencyGraph, ToolCall } from "./dependencyAnalyzer.js";
import type { SynapseLogger } from "./observability.js";

/** A set of tool calls that can execute concurrently. */
export interface Stage {
  index: number;
  calls: ToolCall[];
}

/** The full ordered execution plan produced by the Planner. */
export interface ExecutionPlan {
  stages: Stage[];
  totalCalls: number;
  /** Average calls per stage — higher means more parallelism. */
  parallelism: number;
  /** Sequence of call ids forming the longest dependency chain. */
  criticalPath: string[];
}

/** Builds an ExecutionPlan from a DependencyGraph. */
export class Planner {
  constructor(private logger?: SynapseLogger) {}

  plan(graph: DependencyGraph): ExecutionPlan {
    // Compute in-degree for each node
    const inDegree = new Map<string, number>();
    for (const [id, deps] of graph.edges) {
      inDegree.set(id, deps.size);
    }

    // Start queue with all zero-in-degree nodes
    const queue: string[] = [];
    for (const [id, deg] of inDegree) {
      if (deg === 0) queue.push(id);
    }

    const stages: Stage[] = [];
    let processed = 0;

    while (queue.length > 0) {
      // Drain the current level into a Stage
      const levelSize = queue.length;
      const stage: Stage = { index: stages.length, calls: [] };

      for (let i = 0; i < levelSize; i++) {
        const id = queue.shift()!;
        stage.calls.push(graph.nodes.get(id)!);
        processed++;

        // Reduce in-degree of dependents
        for (const dependentId of graph.revEdges.get(id) ?? []) {
          const newDeg = (inDegree.get(dependentId) ?? 0) - 1;
          inDegree.set(dependentId, newDeg);
          if (newDeg === 0) {
            queue.push(dependentId);
          }
        }
      }

      stages.push(stage);
    }

    if (processed !== graph.nodes.size) {
      throw new Error(
        `Cycle detected during planning — ${processed}/${graph.nodes.size} nodes processed.`
      );
    }

    const totalCalls = graph.nodes.size;
    const parallelism = stages.length > 0 ? totalCalls / stages.length : 0;
    const criticalPath = this.computeCriticalPath(graph);
    const estimatedSpeedup = criticalPath.length > 0 ? totalCalls / criticalPath.length : 1;

    this.logger?.log("dag_planned", {
      parallelCount: Math.max(0, ...stages.map((stage) => stage.calls.length)),
      stageCount: stages.length,
      criticalPath,
      estimatedSpeedup,
    });

    return { stages, totalCalls, parallelism, criticalPath };
  }

  private computeCriticalPath(graph: DependencyGraph): string[] {
    const memo = new Map<string, string[]>();

    const longest = (id: string): string[] => {
      if (memo.has(id)) return memo.get(id)!;
      const preds = graph.edges.get(id) ?? new Set();
      if (preds.size === 0) {
        memo.set(id, [id]);
        return [id];
      }
      let best: string[] = [];
      for (const pred of preds) {
        const path = longest(pred);
        if (path.length > best.length) best = path;
      }
      const result = [...best, id];
      memo.set(id, result);
      return result;
    };

    let globalBest: string[] = [];
    for (const id of graph.nodes.keys()) {
      const p = longest(id);
      if (p.length > globalBest.length) globalBest = p;
    }
    return globalBest;
  }
}
